File: bot.js
#!/usr/bin/env node
/**
 * Quant Trading Bot — Market-Making with Inventory Skew
 * =====================================================
 * Uses Puppeteer to bypass Cloudflare TLS restrictions.
 * All API calls are routed through a headless Chrome browser.
 *
 * Usage:
 *   node bot.js              # Run the bot
 *   node bot.js --dry-run    # Show quotes without placing orders
 */

const puppeteer = require("puppeteer-core");

// ═══════════════════════════════════════════════════════════════
// Configuration
// ═══════════════════════════════════════════════════════════════
const CONFIG = {
    baseUrl: "https://quant.devclub.in/backend/api",
    wsUrl: "wss://quant.devclub.in/backend/ws",
    username: "B23",
    password: "7849",

    stocks: ["Pepsi", "Coca-Cola", "Fanta"],
    maxPosition: 50,
    minPosition: -50,

    // Strategy
    spread: 0.30,          // Base spread around mid-price ($)
    orderSize: 5,          // Default order quantity per side
    inventorySkew: 0.02,   // Price skew per unit of inventory ($)
    refreshInterval: 4000, // ms between strategy cycles
    chromePath: process.env.PUPPETEER_EXECUTABLE_PATH || "/usr/bin/chromium",
};

const DRY_RUN = process.argv.includes("--dry-run");

// ═══════════════════════════════════════════════════════════════
// API Client (all calls go through Puppeteer's page.evaluate)
// ═══════════════════════════════════════════════════════════════
class ApiClient {
    constructor() {
        this.browser = null;
        this.page = null;
        this.token = null;
    }

    async init() {
        this.browser = await puppeteer.launch({
            executablePath: CONFIG.chromePath,
            headless: true,
            args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-gpu"],
        });
        this.page = await this.browser.newPage();
        // Navigate to the site first so fetch calls share the same origin
        await this.page.goto("https://quant.devclub.in", { waitUntil: "domcontentloaded", timeout: 15000 });
        log("Browser initialized");
    }

    async close() {
        if (this.browser) await this.browser.close();
    }

    /** Execute a fetch call inside the browser context */
    async apiFetch(path, options = {}) {
        const url = `${CONFIG.baseUrl}${path}`;
        const token = this.token;
        return this.page.evaluate(
            async ({ url, options, token }) => {
                const headers = { ...(options.headers || {}) };
                if (token) headers["Authorization"] = `Bearer ${token}`;
                if (options.json) headers["Content-Type"] = "application/json";
                const resp = await fetch(url, {
                    method: options.method || "GET",
                    headers,
                    body: options.json ? JSON.stringify(options.json) : options.body,
                });
                const data = await resp.json();
                return { status: resp.status, data };
            },
            { url, options, token }
        );
    }

    async login() {
        const formBody = `username=${encodeURIComponent(CONFIG.username)}&password=${encodeURIComponent(CONFIG.password)}`;
        const result = await this.page.evaluate(
            async ({ url, formBody }) => {
                const resp = await fetch(url, {
                    method: "POST",
                    headers: { "Content-Type": "application/x-www-form-urlencoded" },
                    body: formBody,
                });
                return { status: resp.status, data: await resp.json() };
            },
            { url: `${CONFIG.baseUrl}/auth/login`, formBody }
        );

        if (result.status !== 200) throw new Error(`Login failed: ${JSON.stringify(result)}`);
        this.token = result.data.access_token;
        log(`Logged in as ${CONFIG.username}`);
        return this.token;
    }

    async getStocks() {
        return (await this.apiFetch("/market/stocks")).data;
    }

    async getPrice(stock) {
        return (await this.apiFetch(`/market/${encodeURIComponent(stock)}/current-price`)).data;
    }

    async getOrderbook(stock) {
        return (await this.apiFetch(`/market/${encodeURIComponent(stock)}/orderbook`)).data;
    }

    async getPortfolio() {
        return (await this.apiFetch("/portfolio")).data;
    }

    async placeOrder(stock, side, quantity, price) {
        const payload = {
            stock_symbol: stock,
            order_type: side,
            quantity: Math.floor(quantity),
            price: Math.round(price * 100) / 100,
        };
        const result = await this.apiFetch("/orders", { method: "POST", json: payload });
        log(
            `ORDER ${side.toUpperCase()} ${stock} x${quantity} @ $${price.toFixed(2)} → id=${result.data.id} status=${result.data.status}`
        );
        return result.data;
    }
}

// ═══════════════════════════════════════════════════════════════
// Market-Making Strategy
// ═══════════════════════════════════════════════════════════════
class MarketMaker {
    constructor(client) {
        this.client = client;
        this.spread = CONFIG.spread;
        this.orderSize = CONFIG.orderSize;
        this.skew = CONFIG.inventorySkew;
    }

    computeQuotes(stock, orderbook, currentPos) {
        const bids = orderbook.bids || [];
        const asks = orderbook.asks || [];
        let mid;

        if (bids.length > 0 && asks.length > 0) {
            const bestBid = parseFloat(bids[0].price);
            const bestAsk = parseFloat(asks[0].price);
            mid = (bestBid + bestAsk) / 2;
        } else if (bids.length > 0) {
            mid = parseFloat(bids[0].price) + this.spread / 2;
        } else if (asks.length > 0) {
            mid = parseFloat(asks[0].price) - this.spread / 2;
        } else {
            return []; // No market data, skip
        }

        // Skew based on inventory
        const inventoryOffset = currentPos * this.skew;
        const adjustedMid = mid - inventoryOffset;

        const buyPrice = Math.round((adjustedMid - this.spread / 2) * 100) / 100;
        const sellPrice = Math.round((adjustedMid + this.spread / 2) * 100) / 100;

        // Clamp order sizes to position limits
        const buyQty = Math.min(this.orderSize, CONFIG.maxPosition - currentPos);
        const sellQty = Math.min(this.orderSize, currentPos - CONFIG.minPosition);

        const orders = [];
        if (buyQty > 0) orders.push({ side: "buy", price: buyPrice, qty: buyQty });
        if (sellQty > 0) orders.push({ side: "sell", price: sellPrice, qty: sellQty });

        return orders;
    }

    async execute(stocks) {
        const portfolio = await this.client.getPortfolio();
        const cash = portfolio.cash_balance || 0;
        const positions = {};
        for (const pos of portfolio.positions || []) {
            positions[pos.stock_symbol] = pos.quantity || 0;
        }

        log(`Cash: $${cash.toFixed(2)} | Positions: ${JSON.stringify(positions)}`);

        const placed = [];
        for (const stock of stocks) {
            const currentPos = positions[stock] || 0;
            try {
                const orderbook = await this.client.getOrderbook(stock);
                const quotes = this.computeQuotes(stock, orderbook, currentPos);

                for (const q of quotes) {
                    if (DRY_RUN) {
                        log(`[DRY] Would ${q.side.toUpperCase()} ${stock} x${q.qty} @ $${q.price.toFixed(2)}`);
                    } else {
                        try {
                            const result = await this.client.placeOrder(stock, q.side, q.qty, q.price);
                            placed.push(result);
                        } catch (e) {
                            log(`Order failed ${q.side} ${stock}: ${e.message}`, "ERROR");
                        }
                    }
                }
            } catch (e) {
                log(`Orderbook failed for ${stock}: ${e.message}`, "WARN");
            }
        }
        return placed;
    }
}

// ═══════════════════════════════════════════════════════════════
// Bot Runner
// ═══════════════════════════════════════════════════════════════
function log(msg, level = "INFO") {
    const ts = new Date().toISOString().replace("T", " ").substring(0, 19);
    console.log(`${ts} [${level}] ${msg}`);
}

function printPortfolio(portfolio) {
    const cash = portfolio.cash_balance || 0;
    const pnl = portfolio.pnl || 0;
    const prices = portfolio.current_prices || {};

    log("═".repeat(60));
    log(`  PORTFOLIO  │  Cash: $${cash.toFixed(2)}  │  P&L: $${pnl.toFixed(2)}`);
    log("─".repeat(60));
    for (const pos of portfolio.positions || []) {
        const sym = pos.stock_symbol;
        const qty = pos.quantity || 0;
        const price = prices[sym] || "?";
        log(`  ${sym.padEnd(12)} │  ${(qty >= 0 ? "+" : "") + qty} shares  │  price: $${price}`);
    }
    log("═".repeat(60));
}

async function main() {
    log("═".repeat(60));
    log("  QUANT TRADING BOT — Market Maker");
    log(`  Stocks: ${CONFIG.stocks.join(", ")}`);
    log(`  Spread: $${CONFIG.spread.toFixed(2)}  |  Order Size: ${CONFIG.orderSize}`);
    log(`  Refresh: ${CONFIG.refreshInterval / 1000}s  |  Dry Run: ${DRY_RUN}`);
    log("═".repeat(60));

    const client = new ApiClient();
    await client.init();
    await client.login();

    const strategy = new MarketMaker(client);

    // Show initial portfolio
    const portfolio = await client.getPortfolio();
    printPortfolio(portfolio);

    let cycle = 0;
    let running = true;

    process.on("SIGINT", () => {
        log("Shutdown signal received...");
        running = false;
    });
    process.on("SIGTERM", () => {
        log("Shutdown signal received...");
        running = false;
    });

    while (running) {
        cycle++;
        log(`─── Cycle ${cycle} ───`);

        try {
            const orders = await strategy.execute(CONFIG.stocks);
            if (!DRY_RUN) log(`Placed ${orders.length} orders this cycle`);

            // Print portfolio every 5 cycles
            if (cycle % 5 === 0) {
                try {
                    const p = await client.getPortfolio();
                    printPortfolio(p);
                } catch (e) {
                    log(`Portfolio fetch failed: ${e.message}`, "WARN");
                }
            }
        } catch (e) {
            log(`Cycle ${cycle} error: ${e.message}`, "ERROR");
        }

        // Re-login every 20 cycles to refresh token
        if (cycle % 20 === 0) {
            try {
                await client.login();
            } catch (e) {
                log(`Re-login failed: ${e.message}`, "ERROR");
            }
        }

        await new Promise((r) => setTimeout(r, CONFIG.refreshInterval));
    }

    log("Shutting down...");
    await client.close();
    log("Bot stopped.");
}

main().catch((e) => {
    console.error("Fatal:", e);
    process.exit(1);
});

File: bot.py
#!/usr/bin/env python3
"""
Quant Trading Bot — Market-Making with Inventory Skew
=====================================================
Connects to the DevClub Quant Trading Simulation via REST + WebSocket
and runs a market-making strategy on all available stocks.

Usage:
    python bot.py              # Run the bot
    python bot.py --dry-run    # Show what orders would be placed, don't execute
"""

import sys
import json
import time
import socket
import signal
import logging
import threading
import argparse

import websocket

import config
from api_client import QuantClient
from strategy import AdaptiveStrategy

logging.basicConfig(level=config.LOG_LEVEL, format=config.LOG_FORMAT)
log = logging.getLogger("bot")

shutdown = threading.Event()


def on_shutdown(signum, frame):
    log.info("Shutdown signal received, stopping...")
    shutdown.set()


signal.signal(signal.SIGINT, on_shutdown)
signal.signal(signal.SIGTERM, on_shutdown)


# Force IPv4 for websocket-client as well
_orig_create_connection = websocket._http.HAVE_CONTEXTVAR = False
_orig_socket_create = socket.create_connection


def _ipv4_create_connection(address, *args, **kwargs):
    host, port = address
    # Resolve to IPv4 only
    for res in socket.getaddrinfo(host, port, socket.AF_INET, socket.SOCK_STREAM):
        af, socktype, proto, canonname, sa = res
        try:
            return _orig_socket_create(sa, *args, **kwargs)
        except socket.error:
            continue
    return _orig_socket_create(address, *args, **kwargs)


socket.create_connection = _ipv4_create_connection


class TradingBot:
    def __init__(self, dry_run=False):
        self.client = QuantClient()
        self.strategy = AdaptiveStrategy(self.client)
        self.dry_run = dry_run
        self.ws = None
        self.ws_thread = None

    def print_portfolio(self):
        portfolio = self.client.get_portfolio()
        cash = portfolio.get("cash_balance", 0)
        pnl = portfolio.get("pnl", 0)
        positions = portfolio.get("positions", [])
        current_prices = portfolio.get("current_prices", {})

        log.info("═" * 60)
        log.info("  PORTFOLIO  │  Cash: $%.2f  │  P&L: $%.2f", cash, pnl)
        log.info("─" * 60)
        for sym, qty in positions.items():
            cur_price = current_prices.get(sym, "?")
            log.info("  %-12s │  %+4d shares  │  price: $%s", sym, qty, cur_price)
        log.info("═" * 60)

    def start_websocket(self):
        token = self.client.token or self.client.login()
        ws_url = f"{config.WS_URL}?token={token}"

        def on_message(ws, message):
            try:
                data = json.loads(message)
                msg_type = data.get("type", "unknown")
                if msg_type == "price_update":
                    log.debug("WS price: %s", data)
                elif msg_type in ("order_update", "trade"):
                    log.info("WS %s: %s", msg_type, data)
                else:
                    log.debug("WS: %s", data)
            except json.JSONDecodeError:
                log.debug("WS raw: %s", message)

        def on_error(ws, error):
            log.warning("WS error: %s", error)

        def on_close(ws, close_status, close_msg):
            log.info("WS closed: %s %s", close_status, close_msg)

        def on_open(ws):
            log.info("WebSocket connected")

        self.ws = websocket.WebSocketApp(
            ws_url,
            on_message=on_message,
            on_error=on_error,
            on_close=on_close,
            on_open=on_open,
        )
        self.ws_thread = threading.Thread(
            target=self.ws.run_forever, daemon=True, name="ws"
        )
        self.ws_thread.start()

    def run(self):
        log.info("=" * 60)
        log.info("  QUANT TRADING BOT — Market Maker")
        log.info("  Stocks: %s", ", ".join(config.STOCKS))
        log.info("  Spread: $%.2f  |  Order Size: %d", config.SPREAD, config.ORDER_SIZE)
        log.info("  Refresh: %ds  |  Dry Run: %s", config.REFRESH_INTERVAL, self.dry_run)
        log.info("=" * 60)

        self.client.login()
        self.print_portfolio()

        if not self.dry_run:
            try:
                self.start_websocket()
            except Exception as e:
                log.warning("WebSocket unavailable, using REST polling: %s", e)

        cycle = 0
        while not shutdown.is_set():
            cycle += 1
            log.info("─── Cycle %d ───", cycle)

            try:
                if self.dry_run:
                    self._dry_run_cycle()
                else:
                    orders = self.strategy.execute()
                    log.info("Placed %d orders this cycle", len(orders))

                if cycle % 5 == 0:
                    self.print_portfolio()

                # Re-login every 50 cycles to refresh token
                if cycle % 50 == 0:
                    self.client.login()

            except Exception as e:
                log.error("Cycle %d error: %s", cycle, e)

            shutdown.wait(timeout=config.REFRESH_INTERVAL)

        log.info("Bot stopped.")

    def _dry_run_cycle(self):
        cash, positions = self.strategy.get_positions()
        for stock in config.STOCKS:
            current_pos = positions.get(stock, 0)
            try:
                signals = self.strategy.analyze_stock(stock)
                if signals is None:
                    continue
                orders = self.strategy.generate_orders(signals, current_pos)
                for side, price, qty in orders:
                    log.info(
                        "[DRY] Would %s %s x%d @ $%.2f", side.upper(), stock, qty, price
                    )
            except Exception as e:
                log.warning("[DRY] %s error: %s", stock, e)


import getpass

def main():
    parser = argparse.ArgumentParser(description="Quant Trading Bot")
    parser.add_argument("--dry-run", action="store_true", help="Simulate without placing orders")
    args = parser.parse_args()

    if not config.USERNAME:
        config.USERNAME = input("Username: ")
    if not config.PASSWORD:
        config.PASSWORD = getpass.getpass("Password: ")

    bot = TradingBot(dry_run=args.dry_run)
    bot.run()


if __name__ == "__main__":
    main()
